contracts/schemes/Competition.sol

```diff
@@ -0,0 +1,349 @@
+        uint256 _reputation
+    );
+
+    event SnapshotBlock(
+        bytes32 indexed _proposalId,
+        uint256 _snapshotBlock
+    );
+
+    // A struct holding the data for a competition proposal
+    struct Proposal {
+        uint256 numberOfWinners;
+        uint256[] rewardSplit;
+        uint256 startTime;
+        uint256 votingStartTime;
+        uint256 endTime;
+        uint256 numberOfVotesPerVoters;
+        address payable contributionRewardExt;
+        uint256 snapshotBlock;
+        //contributionRewardExt
+        uint256 reputationReward;
+        uint256 ethReward;
+        uint256 nativeTokenReward;
+        uint256 externalTokenReward;
+
+        uint256[] topSuggestionsOrdered;
+        //mapping from suggestions totalVotes to the number of suggestions with the same totalVotes.
+        mapping(uint256=>uint256) tiesSuggestions;
+        mapping(address=>uint256) votesPerVoter;
+    }
+
+    struct Suggestion {
+        uint256 totalVotes;
+        bytes32 proposalId;
+        address payable suggester;
+        mapping(address=>uint256) votes;
+    }
+
+    //mapping from proposalID to Proposal
+    mapping(bytes32=>Proposal) public proposals;
+    //mapping from suggestionId to Suggestion
+    mapping(uint256=>Suggestion) public suggestions;
+    uint256 public suggestionsCounter;
+
+    /**
+    * @dev Submit a competion proposal
+    * @param _descriptionHash A hash of the proposal's description
+    * @param _reputationChange - Amount of reputation change requested .Can be negative.
+    * @param _rewards rewards array:
+    *         rewards[0] - Amount of tokens requested per period
+    *         rewards[1] - Amount of ETH requested per period
+    *         rewards[2] - Amount of external tokens requested per period
+    * @param _externalToken Address of external token, if reward is requested there
+    * @param _rewardSplit an array of precentages which specify how to split the rewards
+    *         between the winning suggestions
+    * @param _competitionParams competition parameters :
+    *         _competitionParams[0] - competition startTime
+    *         _competitionParams[1] - _votingStartTime competition voting start time
+    *         _competitionParams[2] - _endTime competition end time
+    *         _competitionParams[3] - _numberOfVotesPerVoters on how many suggestions a voter can vote
+    * @param _contributionRewardExt the contributionRewardExt scheme which
+    *        manage and allocate the rewards for the competition.
+    * @return proposalId the proposal id.
+    */
+    function proposeCompetition(
+            string calldata _descriptionHash,
+            int256 _reputationChange,
+            uint[3] calldata _rewards,
+            IERC20 _externalToken,
+            uint256[] calldata _rewardSplit,
+            uint256[4] calldata _competitionParams,
+            address payable _contributionRewardExt //address of the contract to redeem from.
```

Comment:
This is an attack vector with multiple possible exploits. This parameter should probably be part of the `initialize()` method.